### ISSUE TITLE
panic: unresolved function is a placeholder node, but Type was called

### ISSUE THREAD

**Issue by CaptainStabs** [dolthub/dolt#3247]
Version 0.39.0, on hospital-price-transparency-v3
```
dolt sql -q "select code from prices where codein ('21743', '21742') and cms_certification_num in ('010033')
panic: unresolved function is a placeholder node, but Type was called

goroutine 1 [running]:
github.com/dolthub/go-mysql-server/sql/expression.(*UnresolvedFunction).Type(0x1c3663a)
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/expression/unresolved.go:260 +0x27
github.com/dolthub/go-mysql-server/sql/analyzer.fds({0x238fb88?, 0xc012f60a80?})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/analyzer/validation_rules.go:815 +0x52
github.com/dolthub/go-mysql-server/sql/analyzer.logicalSem({{0x238fb88?, 0xc012f60a80?}, {0x238f4a8?, 0xc012f64fa0?}})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/analyzer/validation_rules.go:798 +0x3b
github.com/dolthub/go-mysql-server/sql/analyzer.validateSem({0x238eff8?, 0xc012f64fc0?})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/analyzer/validation_rules.go:785 +0x55
github.com/dolthub/go-mysql-server/sql/analyzer.validateExprSem.func1({0x238eff8?, 0xc012f64fc0?})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/analyzer/validation_rules.go:771 +0x2c
github.com/dolthub/go-mysql-server/sql/transform.exprInspector.Visit(0xc012f5f4c0, {0x238eff8?, 0xc012f64fc0?})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/transform/walk.go:96 +0x31
github.com/dolthub/go-mysql-server/sql.Walk({0x2372c20?, 0xc012f5f4c0?}, {0x238eff8, 0xc012f64fc0})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/walk.go:32 +0x3b
github.com/dolthub/go-mysql-server/sql/transform.WalkExpressions.func1({0x2393b48?, 0xc012f64fe0?})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/transform/walk.go:68 +0xa5
github.com/dolthub/go-mysql-server/sql/transform.inspector.Visit(0xc012f1b950, {0x2393b48?, 0xc012f64fe0?})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/transform/walk.go:49 +0x31
github.com/dolthub/go-mysql-server/sql/transform.Walk({0x2372c60?, 0xc012f1b950?}, {0x2393b48, 0xc012f64fe0})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/transform/walk.go:35 +0x3b
github.com/dolthub/go-mysql-server/sql/transform.Walk({0x2372c60?, 0xc012f1b950?}, {0x23944f8, 0xc000122900})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/transform/walk.go:40 +0xc5
github.com/dolthub/go-mysql-server/sql/transform.Inspect(...)
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/transform/walk.go:60
github.com/dolthub/go-mysql-server/sql/transform.WalkExpressions({0x2372c20?, 0xc012f5f4c0?}, {0x23944f8?, 0xc000122900?})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/transform/walk.go:65 +0x8b
github.com/dolthub/go-mysql-server/sql/transform.InspectExpressions(...)
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/transform/walk.go:90
github.com/dolthub/go-mysql-server/sql/analyzer.validateExprSem(0xc00061b2c0?, 0x2019fc1?, {0x23944f8, 0xc000122900}, 0x0?, 0x0?)
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/analyzer/validation_rules.go:770 +0x8a
github.com/dolthub/go-mysql-server/sql/analyzer.(*Batch).evalOnce(0x23944f8?, 0xc000122900?, 0xc000614e10, {0x23944f8?, 0xc000122900?}, 0xc000806c00?, 0x20b4f10)
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/analyzer/batch.go:120 +0x238
github.com/dolthub/go-mysql-server/sql/analyzer.(*Batch).EvalWithSelector(0xc0004af8f0, 0x320e708?, 0xc000614e10, {0x23944f8, 0xc000122900}, 0x1?, 0xc00060e800?)
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/analyzer/batch.go:66 +0x15f
github.com/dolthub/go-mysql-server/sql/analyzer.(*Batch).Eval(...)
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/analyzer/batch.go:56
github.com/dolthub/go-mysql-server/sql/analyzer.(*Analyzer).analyzeWithSelector(0xc000614e10?, 0x23944f8?, {0x23944f8?, 0xc000122900?}, 0xc0009e4510?, 0x20b4f18?, 0x0?)
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/analyzer/analyzer.go:474 +0x347
github.com/dolthub/go-mysql-server/sql/analyzer.(*Analyzer).Analyze(...)
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/sql/analyzer/analyzer.go:369
github.com/dolthub/go-mysql-server.(*Engine).analyzeQuery(0xc000614eb0, 0x23a44c8?, {0xc000634de0?, 0xbf8ae0?}, {0x0?, 0x0?})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/engine.go:293 +0xa7
github.com/dolthub/go-mysql-server.(*Engine).QueryNodeWithBindings(0xbf8c05?, 0xc00061b180, {0xc000634de0, 0x5f}, {0x0, 0x0}, 0xc000634e40?)
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/engine.go:180 +0x10a
github.com/dolthub/go-mysql-server.(*Engine).QueryWithBindings(...)
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/engine.go:159
github.com/dolthub/go-mysql-server.(*Engine).Query(0xc000634de0?, 0x5f?, {0xc000634de0?, 0x0?})
        /go/pkg/mod/github.com/dolthub/go-mysql-server@v0.11.1-0.20220415151345-1911dee4b070/engine.go:154 +0x26
github.com/dolthub/dolt/go/cmd/dolt/commands/engine.(*SqlEngine).Query(...)
        /src/cmd/dolt/commands/engine/sqlengine.go:182
github.com/dolthub/dolt/go/cmd/dolt/commands.processQuery(0xc00039a780?, {0xc000634de0, 0x5f}, 0xc000122810)
        /src/cmd/dolt/commands/sql.go:1034 +0x4f3
github.com/dolthub/dolt/go/cmd/dolt/commands.runMultiStatementMode(0x23829a0?, 0xc000122810, {0x2371c80?, 0xc000089940}, 0x0)
        /src/cmd/dolt/commands/sql.go:652 +0x2fa
github.com/dolthub/dolt/go/cmd/dolt/commands.execMultiStatements({0x23829a0, 0xc000717620}, 0x61?, 0x8?, {0x2371c80, 0xc000089940}, 0xa0?, {0xc000c2e2a0?, 0x16acfbd?})
        /src/cmd/dolt/commands/sql.go:445 +0x14a
github.com/dolthub/dolt/go/cmd/dolt/commands.queryMode({0x23829a0, 0xc000717620}, 0x2000fde?, 0x4b8?, 0xc0004ae540, {0xc0000960c0, 0x5f}, {0xc000c2e2a0, 0x1e}, 0x0, ...)
        /src/cmd/dolt/commands/sql.go:342 +0x25f
github.com/dolthub/dolt/go/cmd/dolt/commands.SqlCmd.Exec({{0x2000fde?, 0xc012e6b910?}}, {0x23829a0, 0xc000717620}, {0xc000047dc0, 0x8}, {0xc00005c0a0, 0x2, 0x2}, 0xc0003f7ef0)
        /src/cmd/dolt/commands/sql.go:203 +0x82b
github.com/dolthub/dolt/go/cmd/dolt/cli.SubCommandHandler.handleCommand({{0x1fff0d9, 0x4}, {0x2018c29, 0x11}, {0x0, 0x0}, {0xc00062adc0, 0x2b, 0x2b}, 0x0}, ...)
        /src/cmd/dolt/cli/command.go:222 +0x29e
github.com/dolthub/dolt/go/cmd/dolt/cli.SubCommandHandler.Exec({{0x1fff0d9, 0x4}, {0x2018c29, 0x11}, {0x0, 0x0}, {0xc00062adc0, 0x2b, 0x2b}, 0x0}, ...)
        /src/cmd/dolt/cli/command.go:179 +0x47f
main.runMain()
        /src/cmd/dolt/dolt.go:345 +0x6a5
main.main()
        /src/cmd/dolt/dolt.go:131 +0x19
```

**Comment by zachmu** [comment]
We should never panic on this kind of error. It should have been caught much earlier in analysis (`codein` is not a valid function)

**Comment by timsehn** [comment]
This is pretty urgent because this will crash a hosted instance.

**Comment by fulghum** [comment]
Here's a simplified repro case that results in the same panic: `select * from dual where foo() and true;`

I'll run this through the debugger and see what's going on.

**Comment by fulghum** [comment]
I was able to repro this with dolt version 0.38.0, but couldn't repro with the latest tip of main in dolt or go-mysql-server. I also verified that this panic doesn't repro with the latest released version of dolt, 0.40.0.

**Comment by fulghum** [comment]
This was fixed in: https://github.com/dolthub/go-mysql-server/commit/324e43b896d344f745be0b52935640335e9fd546

I opened up a PR to add a regression test case: https://github.com/dolthub/go-mysql-server/pull/1000
